Sources/AWSSDKSwiftCore/AWSClient+Paginate.swift

```diff
@@ -0,0 +1,116 @@
+//
+//  AWSClient+Paginate.swift
+//  AWSSDKSwift
+//
+//  Created by Adam Fowler on 2020/01/19.
+//
+//
+import NIO
+
+/// protocol for all AWSShapes that can be paginated.
+/// Adds an initialiser that does a copy but inserts a new string based pagination token
+public protocol AWSPaginateStringToken: AWSShape {
+    init(_ original: Self, token: String)
+}
+
+/// protocol for all AWSShapes that can be paginated.
+/// Adds an initialiser that does a copy but inserts a new integer based pagination token
+public protocol AWSPaginateIntToken: AWSShape {
+    init(_ original: Self, token: Int)
```

Comment:
I'm not a big fan of the naming in this case. I doesn't feel swifty: 

URL is an example we can have a look at...
Creating a new instance by changing one property: `appendingPathComponent(_:)`

This is why I propose:
```
func usingPaginationToken(_:) -> Self
```

https://developer.apple.com/documentation/foundation/url/1780239-appendingpathcomponent